### PR TITLE
Add repair job that will ensure that secret and passwordsalt are set

### LINKED ISSUE #34780: [Bug]: passwordsalt migration missing
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Upgrading to 25 with a config.php that has no or an empty passwordsalt results in an error message. 
There is no clear solution on how to introduce a passwordsalt setting to an older setup.

Adiddional description on discord: https://help.nextcloud.com/t/passwordsalt-missing-from-config-php/148081/2

### Steps to reproduce

1.install 24
2.use a config.php without a passwordsalt
3.upgrade to 25

### Expected behavior

A warning before the upgrade or migration/guide to use salted password hashes.

Possible questions:
Does setting passwordsalt make passwords unusable?
What must be done to migrate this setting?

### Installation method

Community Docker image

### Operating system

Other

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "overwrite.cli.url": "https:\/\/nextcloud.domain.at",
        "enable_previews": false,
        "overwriteprotocol": "https",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "25.0.0.18",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "theme": "",
        "forcessl": true,
        "trusted_domains": [
            "owncloud.domain.at",
            "nextcloud.domain.at"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "overwritehost": "nextcloud.domain.at",
        "share_folder": "\/Shared",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 0,
        "updater.release.channel": "stable",
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": true,
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "default_phone_region": "AT",
        "mysql.utf8mb4": true
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - calendar: 4.0.1
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - comments: 1.15.0
  - contacts: 5.0.1
  - contactsinteraction: 1.6.0
  - dashboard: 7.5.0
  - dav: 1.24.0
  - deck: 1.8.0
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_external: 1.17.0
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - nextcloud_announcements: 1.14.0
  - notes: 4.6.0
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 2.0.0
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - recommendations: 1.4.0
  - related_resources: 1.0.1
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - spreed: 15.0.0
  - support: 1.8.0
  - survey_client: 1.13.0
  - systemtags: 1.15.0
  - tasks: 0.14.5
  - text: 3.6.0
  - theming: 2.0.0
  - twofactor_backupcodes: 1.14.0
  - updatenotification: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflowengine: 2.7.0
Disabled:
  - bruteforcesettings: 2.4.0
  - encryption
  - files_texteditor: 2.14.0
  - suspicious_login
  - twofactor_totp
  - user_ldap
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### LINKED ISSUE #35347: [Bug]: NC25 - The required secret config variable is not configured in the config.php file
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

After upgrade to NC 25.0.1 - login page is not available due to verification of configuration parameters. Please note that my installation is very old and there is no secret parameter defined in configuration. Untill version 24.0.7 - this was perfectly fine.

Workaround for my installation (point 1 from "expected behavior") is described [here](https://help.nextcloud.com/t/upgrade-from-24-0-7-to-25-0-1-missing-secret-in-config/150208/6?u=piknew).

### Steps to reproduce

1. have installation of NC 24.x without secret defined in config.
2. upgrade to NC 25.0.1 (by web). Logout from NC 25.0.1 (this was
3. see login page - there is validation message **_The required secret config variable is not configured in the config.php file_**.
4. there is no way to proceed with login to NC. NC Clients are also not able to login.

### Expected behavior

System shall (options):
1. Not require secret to be defined in config (this may break some features of NC25 which I maybe not aware of). Or define "empty_secret": true in config to forcibly override the check.
2. Prevent upgrade from execution.
3. Provide mechanism to add/generate secret and update of all dependent data.

### Installation method

Community Web installer on a VPS or web space

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "25.0.1.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "log_type": "syslog",
        "logfile": "\/share\/spool\/nextcloud\/data\/nextcloud.log",
        "loglevel": 2,
        "cron_log": true,
        "syslog_tag": "nextcloud",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "maintenance": false,
        "theme": "",
        "updater.release.channel": "stable",
        "mysql.utf8mb4": true,
        "mail_sendmailmode": "smtp",
        "defaultapp": "files",
        "default_phone_region": "pl",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "htaccess.RewriteBase": "\/"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - comments: 1.15.0
  - contactsinteraction: 1.6.0
  - dashboard: 7.5.0
  - dav: 1.24.0
  - drawio: 1.0.5
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_external: 1.17.0
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - groupfolders: 13.0.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - maps: 0.2.1
  - metadata: 0.17.0
  - nextcloud_announcements: 1.14.0
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 2.0.0
  - previewgenerator: 5.1.1
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - recommendations: 1.4.0
  - related_resources: 1.0.3
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - support: 1.8.0
  - survey_client: 1.13.0
  - systemtags: 1.15.0
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_backupcodes: 1.14.0
  - updatenotification: 1.15.0
  - user_ldap: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflowengine: 2.7.0
Disabled:
  - bruteforcesettings: 1.0.2
  - encryption: 2.9.0
  - suspicious_login
  - twofactor_totp
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by CarlSchwan**
* Resolves: #35347 
* Resolves: #34780 
 
## Summary

Ensure that the secret and passwordsalt config are set

### CODE REVIEW

File: lib/private/Repair/NC25/AddMissingSecretJob.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		$this->random = $random;
+	}
+
+	public function getName(): string {
+		return 'Add possibly missing system config';
+	}
+
+	public function run(IOutput $output): void {
+		$passwordSalt = $this->config->getSystemValue('passwordsalt', null);
+		if ($passwordSalt === null) {
```
Comment (szaimen) [inline lib/private/Repair/NC25/AddMissingSecretJob.php:50]: What if passwordsalt is set to an empty string or some other incorrect value?

**Comment by ChristophWurst** [comment]
The secret is generated during installation, isn't it? And for the ancient installations we already had a repair step somewhere. Was that dropped?

If the secret was *lost*, then any encrypted values where the instance secret was used as password will not be recoverable. We have seen that a few times. So with the automatic regeneration we'll possibly cause that problem.

So I'm thinking if the automatic recreation is really the best approach. The alternative could be to show a big error/warning on the setup page. Explain the situation and admins can first check if they have the secret in a config backup.

Just an idea.

Otherwise :+1:

File: lib/private/Repair/NC25/AddMissingSecretJob.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			try {
+				$this->config->setSystemValue('passwordsalt', $this->random->generate(30));
+			} catch (HintException $e) {
+				$output->warning("passwordsalt is missing from your config.php and your config.php is read only. Please fix it manually.");
+			}
+		}
+
+		$secret = $this->config->getSystemValue('secret', null);
```
Comment (szaimen) [inline lib/private/Repair/NC25/AddMissingSecretJob.php:58]: What if secret is set to an empty string or some other incorrect value?

File: lib/private/Repair/NC25/AddMissingSecretJob.php (hunk 1 above)
Comment (szaimen) [reply lib/private/Repair/NC25/AddMissingSecretJob.php:50]: (Should be checked one line below)

File: lib/private/Repair/NC25/AddMissingSecretJob.php (hunk 2 above)
Comment (szaimen) [reply lib/private/Repair/NC25/AddMissingSecretJob.php:58]: (Should be checked one line below)

**Comment by szaimen** [comment]
> So I'm thinking if the automatic recreation is really the best approach. The alternative could be to show a big error/warning on the setup page. Explain the situation and admins can first check if they have the secret in a config backup.

@CarlSchwan @juliushaertl wasnt that the idea in the first place?

**Comment by CarlSchwan** [comment]
> If the secret was _lost_, then any encrypted values where the instance secret was used as password will not be recoverable. We have seen that a few times. So with the automatic regeneration we'll possibly cause that problem.

We already have a fallback in place where if a password hash or encrypted password was previously encrypted with an empty secret, we check first with the current secret and then with an empty secret. See 81f8719cc0fdc2fc25a00ac1412ffc63cd62eabf

**Comment by ChristophWurst** [comment]
Right. I also now saw the referenced ticket #34780.

**Review by ChristophWurst — APPROVED** [review]
:crossed_fingers:

File: lib/private/Repair/NC25/AddMissingSecretJob.php (hunk 2 above)
Comment (CarlSchwan) [reply lib/private/Repair/NC25/AddMissingSecretJob.php:55]: I am now also checking for an empty string, any other value at all would not have worked even before 25

File: lib/private/Repair/NC25/AddMissingSecretJob.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2022 Thomas Citharel <nextcloud@tcit.fr>
```
[the remark this answers is not in the document]
Comment (tcitworld) [reply lib/private/Repair/NC25/AddMissingSecretJob.php]: Well hello there

**Review by PVince81 — APPROVED** [review]
:+1:

**Pull request merged**